### DISCUSSION TITLE
utorrent-block-xunlei-v2.2

### DISCUSSION THREAD

**Discussion by ShenHongFei** [ShenHongFei/utorrent-block-xunlei#25] [General]
## 使用方法
1. 确保 uTorrent 是 3.0 以上版本（不支持 2.x 版本），并开启 WebUI (网页界面)  
1.1 打开 uTorrent 设置 > 高级 > 网页界面  
1.2 选上 "启用网页界面", 并在下方填写用户和密码, 记下来, 作为命令行 `--username` 和 `--password` 参数的值  
2. 在 "uTorrent 设置 > 连接 > 监听端口 > 传入连接所使用的端口" 中查看设置的端口号，记下来，作为命令行 `--port` 参数的值
3. 在 uTorrent 数据目录中（路径可能是 `C:/Users/<用户名>/AppData/Roaming/uTorrent/ipfilter.dat` ，该目录下应该有 resume.dat 文件存在）保证 ipfilter.dat 文件存在（若不存在则新建空白 ipfilter.dat 文件），脚本会在原有 ipfilter.dat 文件内容之后添加被屏蔽的 IP，不影响已有内容及其功能
4. 在 releases 中下载 `utorrent-block-xunlei.exe`, 下载地址: [https://github.com/ShenHongFei/utorrent-block-xunlei/releases](https://github.com/ShenHongFei/utorrent-block-xunlei/releases)
5. 查看用法: `utorrent-block-xunlei.exe --help`
```text
Usage: utorrent-block-xunlei --port 50050 --username tom --password 123456 --ipfilter "C:/Users/tom/AppData/Roaming/uTorrent/ipfilter.dat"

Xunlei Blocker of uTorrent

Options:
  -V, --version                output the version number
  --hostname <hostname>        选填, uTorrent WebUI 主机的 hostname, 默认为本机 127.0.0.1, 也可设置远程主机 IP 或域名 (default: "127.0.0.1")
  --port <port>                必填, "uTorrent 设置 > 连接 > 监听端口 > 传入连接所使用的端口" 中设置的端口号
  --username <username>        必填, "uTorrent 设置 > 高级 > 网页界面 > 身份验证 > 用户" 中设置的用户名
  --password <password>        必填, "uTorrent 设置 > 高级 > 网页界面 > 身份验证 > 密码" 中设置的密码
  --ipfilter <ipfilter>        必填, uTorrent 数据目录中 ipfilter.dat 文件的完整路径，如: C:/Users/tom/AppData/Roaming/uTorrent/ipfilter.dat
  --interval <interval>        选填, 检测 peers 的间隔（秒）, 默认每隔 20 秒检测并屏蔽一次 (default: "20")
  --reset-ipfilter <interval>  选填, 在启动时以及每间隔 interval 秒后自动清空 ipfilter.dat 的内容
  -h, --help                   display help for command
```

6. 启动 blocker (修改下参数的值): `utorrent-block-xunlei --port 50050 --username shf --password 123456 --ipfilter "C:/Users/shf/AppData/Roaming/uTorrent/ipfilter.dat"`

7. (可选，不影响屏蔽功能) 在 uTorrent 中开启屏蔽日志，查看被屏蔽的连接请求
    4.1 在 uTorrent 下方的日志面板中点击右键
    4.2 勾选 `记录用户通讯信息 > 记录拦截连接`
    4.3 有被屏蔽的用户连接时可以在 uTorrent 日志面板中看到类似下面的日志
```text
[2018-11-22 19:03:43]  Loaded ipfilter.dat (51 entries)
[2018-11-22 19:03:46]  IpFilter blocked peer 223.81.192.235
[2018-11-22 19:03:49]  IpFilter blocked peer 223.81.192.235
[2018-11-22 19:04:06]  IpFilter blocked peer 223.81.192.235
[2018-11-22 19:04:21]  IpFilter blocked peer 183.25.54.216
[2018-11-22 19:04:46]  IpFilter blocked peer 223.81.192.235
...
```

8. 实时控制，查看状态，REPL
在 `utorrent-block-xunlei` 命令行窗口中输入 `utorrent` 可通过 REPL 查看其属性，调用其方法
如输入 `utorrent.reset_ipfilter()` 可清空屏蔽列表和 ipfilter.dat, `utorrent.print_blockeds()` 可查看被屏蔽的 IP

<hr /><em>This discussion was created from the release <a href='https://github.com/ShenHongFei/utorrent-block-xunlei/releases/tag/v2.2'>utorrent-block-xunlei-v2.2</a>.</em>

**Comment by tonyhsie** [comment]
> 加了，在 release 里面的 v2.2 版本，你可以试用下

感謝，馬上來試試！

**Reply by tonyhsie** [reply]
試了一下，在每一段間隔時間過後，是有看到 "write: C:/Users/xxx/AppData/Roaming/uTorrent/ipfilter.dat" 的訊息
但無法知道是不是有作用

一啟動 utorrent-block-xunlei 後 ipfilter.dat 一直維持檔案大小為 0 的狀態
就算關閉了 utorrent-block-xunlei 之後也不會恢復

不曉得這算不算是 bug

**Reply by ShenHongFei** [reply]
是正常工作的
“ipfilter.dat 一直維持檔案大小為 0 的狀態”是因为没有检测到需要屏蔽的迅雷 IP

**Reply by tonyhsie** [reply]
但在關閉 utorrent-block-xunlei 後，ipfilter 應該要恢復原本的檔案大小才對?

目前我這邊不會恢復，等於是執行了 utorrent-block-xunlei 之後，它就把 ipfilter 給吃掉了一樣
1.0 版並不會這樣

**Reply by ShenHongFei** [reply]
ipfilter.dat 是用来保存需要屏蔽的 IP，没有别的作用，你可以用记事本打开查看里面的内容。
如果需要定时清空屏蔽 IP 的功能，我理解不需要保留运行之前的 IP 屏蔽列表（也就是 ipfilter.dat 的内容）吧

**Reply by SeaHOH** [reply]
> 如果需要定时清空屏蔽 IP 的功能，我理解不需要保留运行之前的 IP 屏蔽列表（也就是 ipfilter.dat 的内容）吧

需要保留，只清除 blocker 自己添加的。我的实践是保留写成范围格式的；单个 IP 的，记录屏蔽时间，到时删除。

**Reply by tonyhsie** [reply]
定時清空屏蔽 IP，指的應該是「清空 utorrent-block-xunlei 執行之後所新增的部分」
而不是清空整個 ipfilter.dat 吧?


例如我現在 utorrent 目錄下有個 15MB 的 ipfilter.dat，執行了  utorrent-block-xunlei 後，它接管，保存在記憶體裡，
然後就清空了 ipfilter.dat

到這裡還沒有什麼問題


如果我關閉了 utorrent-block-xunlei，理論上在 utorrent 目錄下，應該要有原來的 15MB 大小的 ipfilter.dat
但現在 2.2 版的實際結果是，目錄下只剩一個檔案大小為 0 的 ipfilter.dat

「脚本会在原有 ipfilter.dat 文件内容之后添加被屏蔽的 IP，不影响已有内容及其功能」

在 1.0 版時確實是如此，但 2.2 版會把 ipfilter.dat 「永久清空」，這樣的結果好像有點怪異

**Reply by tonyhsie** [reply]
PS.

原 ipfilter.dat 裡的 ip 都是要永久封鎖的，任何時候都不需要清空

**Reply by ShenHongFei** [reply]
好的，那我改下逻辑，定时清除的时候只清空本次运行新增的 IP，保留运行之前 ipfilters 中已有的

**Reply by ShenHongFei** [reply]
@tonyhsie @SeaHOH v2.3 修改了逻辑，可以再试下

**Reply by tonyhsie** [reply]
感謝，晚點再來試試!

**Reply by SeaHOH** [reply]
重置后行为与启动时产生了差异
  - 启动时，会加载原本的 ipfilter 内容
  - 重置后，丢失了 ipfilter 内容，直到正常退出 blocker 后才会恢复

**Reply by ShenHongFei** [reply]
> 重置后行为与启动时产生了差异
> 
> * 启动时，会加载原本的 ipfilter 内容
> * 重置后，丢失了 ipfilter 内容，直到正常退出 blocker 后才会恢复

我试试，如果这样的话那就是 Bug

**Reply by ShenHongFei** [reply]
找到 Bug 了，我修复下

**Reply by ShenHongFei** [reply]
@SeaHOH @tonyhsie v2.4 已修复，再试试吧
